### ISSUE TITLE
Make Scaling in BatchNorm optional 

### ISSUE THREAD

**Issue by fkemeth** [pytorch/pytorch#40087]
## 🚀 Feature

Using the learnable scale γ in batch normalization (`weights` in [_BatchNorm)](https://pytorch.org/docs/1.1.0/_modules/torch/nn/modules/batchnorm.html) should be optional.

## Motivation

If batch normalization is used before a piecewise linear function such as ReLU, the learnable scale γ is practically redundant. It is therefore common to set it equal to 1 in such cases. (See, for example, last paragraph in Section 4 in [https://arxiv.org/ab/1710.05941.pdf](https://arxiv.org/ab/1710.05941.pdf)].

## Pitch

A boolean flag `scale` in `class _BatchNorm(Module)` which sets `self.weights` to ones if `False`.

I am implementing this. If you also think this is useful in general, I can create a PR.

**Comment by SsnL** [comment]
> If batch normalization is used before a piecewise linear function such as ReLU, the learnable scale γ is practically redundant. 

Could you elaborate why this is true?

**Comment by fkemeth** [comment]
There is a brief discussion [here](https://datascience.stackexchange.com/questions/22073/why-is-scale-parameter-on-batch-normalization-not-needed-on-relu), which may be summarized as follows:

Consider batch normalization of an input x

<img src="https://latex.codecogs.com/gif.latex?y=\mathrm{BN(x,\gamma,\beta)}=\frac{x-\mathrm{E}[x]}{\sqrt{\mathrm{Var}[x]+\epsilon}}\cdot\gamma+\beta" /> 
which may as well be expressed as

<img src="https://latex.codecogs.com/gif.latex?y=\gamma\left(\frac{x-\mathrm{E}[x]}{\sqrt{\mathrm{Var}[x]+\epsilon}}+\frac{\beta}{\gamma}\right)=\gamma\cdot\mathrm{BN(x,1,\beta/\gamma)}=\gamma\cdot\hat{x}" /> 

If this is followed by, say, a ReLU function `z=max(0, y)`, then

<img src="https://latex.codecogs.com/gif.latex?z=\mbox{max}(0,\gamma\hat{x})=\gamma\cdot\mbox{max}(0,\hat{x})" /> 

If this is followed by a layer with weights `W`, then

<img src="https://latex.codecogs.com/gif.latex?Wz=W\gamma\mbox{max}(0,\hat{x})=\hat{W}\mbox{max}(0,\hat{x})" /> 

This means that having a learnable scale in the batch normalization layer which is followed by a piecewise linear function and another layer does not add additional expressiveness to the model, since it can be absorped in β and W.

**Comment by SsnL** [comment]
Oh I see. Thanks for the explanation. It confused me because this redundancy also exists if BN is used after non-linearity (before linear layers).

**Comment by malfet** [comment]
Thank you for your suggestion. We would accept a patch that fixes this.

**Comment by jackwayneright** [comment]
I just wanted to note that occasionally it's not just redundant, but in some cases it can be problematic to have the extra scaling. If batch normalization has the ability to apply a scale and is followed by weights that are regularized (e.g., by L2 regularization), the batch normalization can arbitrarily scale the values upwards so that the following weights can be arbitrarily small to avoid the regularization. I've run into the issue previously, though not in PyTorch. And it might not be as concerning of an issue in PyTorch since it appears the standard way to apply weight regularization is at the optimizer rather than on a per-layer basis, so only those who require a per-layer regularization would probably run into this issue.
